### PR TITLE
[DeviceSanitizer] Don't instrument referenced-indirectly functions

### PR DESCRIPTION
**Pull request by wenju-he**
When we create SLM __AsanLaunchInfo and store newly added kernel arg __asan_launch into the SLM, the SLM is loaded in asan report function. If instructions in referenced-indirectly function are instrumented, the report function is called. However, access to SLM in referenced- indirectly function isn't supported in intel-graphics-compiler yet.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp [hunk 1]
```diff
@@ -3363,6 +3363,12 @@ bool AddressSanitizer::instrumentFunction(Function &F,
   if (F.getName().contains("__sycl_service_kernel__"))
     return false;
 
+  // Skip referenced-indirectly function as we may insert access to shared local
+  // memory (SLM) __AsanLaunchInfo in report function and access to SLM in
+  // referenced-indirectly function isn't supported in intel-graphics-compiler.
+  if (TargetTriple.isSPIR() && F.hasFnAttribute("referenced-indirectly"))
+    return false;
+
```
Comment (AllanZyne) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:3371]: I think we can move `if (F.getName().contains("__sycl_service_kernel__"))` into one block.
```suggestion
  // Skip referenced-indirectly function as we may insert access to shared local
  // memory (SLM) __AsanLaunchInfo in report function and access to SLM in
  // referenced-indirectly function isn't supported in intel-graphics-compiler.
  if (TargetTriple.isSPIR()) {
    if (F.getName().contains("__sycl_service_kernel__"))
      return false;
   // Skip referenced-indirectly function as we may insert access to shared local
  // memory (SLM) __AsanLaunchInfo in report function and access to SLM in
  // referenced-indirectly function isn't supported in intel-graphics-compiler.
    if (F.hasFnAttribute("referenced-indirectly"))
      return false;
  }
```

Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: done

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp [hunk 2]
```diff
@@ -3360,8 +3360,16 @@ bool AddressSanitizer::instrumentFunction(Function &F,
   if (F.getLinkage() == GlobalValue::AvailableExternallyLinkage) return false;
   if (!ClDebugFunc.empty() && ClDebugFunc == F.getName()) return false;
   if (F.getName().starts_with("__asan_")) return false;
-  if (F.getName().contains("__sycl_service_kernel__"))
-    return false;
+
+  if (TargetTriple.isSPIR()) {
+    if (F.getName().contains("__sycl_service_kernel__"))
```
Comment (wenju-he) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:3365]: @AllanZyne the name check might not be accurate, what if there is a user function that is xx__sycl_service_kernel__xx?

Comment (AllanZyne) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:3365]: Okay. 
As far as I know, currently this kernel naming style is only used for assert.

llvm/sycl/unittests/assert/assert.cpp:55

**Review by AllanZyne — APPROVED** [review]

**Comment by wenju-he** [comment]
@intel/llvm-gatekeepers please merge, thank you.

**Pull request merged**